src/Raven.Server/Documents/Replication/DocumentReplicationLoader.cs

```diff
@@ -59,30 +59,39 @@ protected override void LoadConfigurations()
         }
 
         //inbound replication source will get it's ReplicationExecuter as well as
-        //the outgoing ones
+        //the outgoing ones		
         public DocumentReplicationExecuter RegisterNewConnectionFrom(
             Guid srcDbId, 
             string srcUrl, 
             string srcDbName, 
-            long lastSentEtag, 
             out bool shouldConnectBack)
         {
-            shouldConnectBack = false;
+            //since this should be done once per destination node, 
+            //the mutext is not likely to be a bottleneck
+            lock (Replications)
+            {
+                shouldConnectBack = false;
 
-            //prevent initiating two connections to the same db
-            if (!_activeConnections.TryAdd(srcDbId))
-                return null;
+                var destination = _replicationDocument.Destinations
+                    .FirstOrDefault(x => x.Url.Equals(srcUrl, StringComparison.OrdinalIgnoreCase) &&
+                                         x.Database.Equals(srcDbName, StringComparison.OrdinalIgnoreCase));
 
-            var destination = _replicationDocument.Destinations
-                .FirstOrDefault(x => x.Url.Equals(srcUrl, StringComparison.OrdinalIgnoreCase) &&
-                                     x.Database.Equals(srcDbName, StringComparison.OrdinalIgnoreCase));
+                if (destination != null)
+                    shouldConnectBack = true;
 
-            if (destination != null)
-                shouldConnectBack = true;
+                var documentReplicationExecuter = new DocumentReplicationExecuter(_database, srcDbId, destination);
+                Replications.Add(documentReplicationExecuter);
+                return documentReplicationExecuter;
+            }
+        }
 
-            var documentReplicationExecuter = new DocumentReplicationExecuter(_database, srcDbId, destination, lastSentEtag);
-            Replications.Add(documentReplicationExecuter);
-            return documentReplicationExecuter;
+        public void HandleConnectionDisconnection(DocumentReplicationExecuter replicationExecuter)
+        {
+            lock (Replications)
+            {
+                replicationExecuter.Dispose();
+                Replications.TryRemove(replicationExecuter);
+            }
```

Comment:
The order of operations here is strange. You first dispose the object, then remove it from the collection.
Other threads might be accessing the `Replications` property and getting a disposed object.
You are better off first removing, then disposing.